### ISSUE TITLE
Rapidly changing the source separation wrecks performance

### ISSUE THREAD

**Issue by samreid** [phetsims/wave-interference#316]
Even before increasing the lattice size, it is possible to make the simulation unusably slow by:

1. Interference
2. Turn sources on
3. Drag separation slider back and forth many times

Discovered during #315

**Comment by samreid** [comment]
You can simulate the same problem by clicking the wave source on and off many many times.  There is a bug in `TemporalMask.prune`.

**Comment by samreid** [comment]
The aforementioned fix corrects a memory leak and allows TemporalMask entries to be pruned when they have left the window.  However, rapidly dragging the source separation slider or pressing the on/off button can still swamp the TemporalMask with deltas.  More than 20 or so kills performance on my Mac Chrome.  The performance is restored in 15 seconds or so when the wave has moved out of the visible region.

@arouinfar and @ariel-phet how do you want to address this problem?  Here are some possibilities:

* Clear the waves when the source separation changes (may also eliminate spurious artifacts, but makes it less user friendly when looking at a changing interference pattern in the wave area or on the screen).
* Only keep the last 5 or so deltas (when on/off or source separation changes).  If you rapidly turn the wave on and off, the more distant waves will disappear.
* Some combination of the preceding ideas
* Your idea here

This issue blocks next round of dev testing, please let me know your thoughts or reach out to me on Slack to discuss further.

**Comment by samreid** [comment]
Also, @arouinfar I wonder if this is what you observed when the performance dropped on the Chromebook you tested.

**Comment by ariel-phet** [comment]
@samreid @arouinfar 

This is just not a pedagogically productive action at all (rapid changing of separation). It seems perhaps keeping only the last 5 or so deltas is a good solution. Basically, I am fine with a solution that allow performance to remain, but if things look a bit odd with you moving the separation slider back and forth rapidly, I am fine with that, since it is not a productive interaction (and it should be a rare interaction).

**Comment by arouinfar** [comment]
@samreid I'm fairly certain this what I was seeing yesterday on the Chromebook. Perhaps I wasn't always changing separation rapidly or often enough to see it happen consistently.

I agree with @ariel-phet. Rapidly scrubbing the separation slider serves no pedagogical purpose, and would be a rare interaction. Keeping the last 5ish deltas also sounds reasonable to me. I do wonder what would happen in the lattice, though. Would part of the lattice clear itself when the 6th delta occurs?

**Comment by samreid** [comment]
Yes, the temporal mask would "forget" the parts furthest from the oscillator.  You can see it in practice on phettest: http://phettest.colorado.edu/wave-interference/wave-interference_en.html?ea&brand=phet

Try the following tests:

Test 1
1. On the Waves screen
2. select the Speaker scene
3. Click the green speaker button 6 times before the wave reaches the right edge

Test 2
1. On the Interference screen
2. Select the Speaker scene
3. Turn on the sources
4. Drag the separation slider (so that the value changes at least 5 times).

Please let me know your thoughts.

**Comment by samreid** [comment]
Also, the proposed "keep only 5 deltas" solution helps performance, but isn't entirely consistent--it clears the temporal mask without clearing the lattice so you can end up with situations where the light screen/chart is lit up without a wave there:

![image](https://user-images.githubusercontent.com/679486/51201734-0c3fc400-18ba-11e9-9621-8a5ea73a74d5.png)

**Comment by ariel-phet** [comment]
@samreid I think given the nature of this issue (rare use case), dealing with the artifacts can be deferred until after 1.0 is published.

**Comment by ariel-phet** [comment]
@samreid @arouinfar what about if we enforce that the source does not emit a wave while being moved, and not until it "settles" (so I slide to a new position, not until I release the slider does it begin to emit a wave again). Would that solve the general artifact issue? It seems seeing what the source is doing while moving is not pedagogically important.

**Comment by samreid** [comment]
I believe that eliminate the numerical artifacts and help with the performance problem, but the user would no longer be able to see the interference pattern "morph" to another pattern.  Should I investigate?

**Comment by samreid** [comment]
The above commits "mute" the sources while the separation is being changed.  This was easy thanks to the pre-existing support in NumberControl, thanks @pixelzoom!

Setting the amplitude to 0 in each frame created odd artifacts at the wave sources, and would have required more complex logic to suppress deltas from the temporal mask.  Instead I tried the simpler solution of just skipping setting point source values altogether.  There is still an artifact when the wave value immediately stops oscillating, but this solves the problem when continuously changing the source separation, and also prevents those values from entering the list of deltas.

@arouinfar can you please test in master?

**Comment by arouinfar** [comment]
@samreid I think things are generally looking a bit better, I do find the artifacts in the lattice to be pretty distracting.

I'm not so sure the sim is behaving as proposed in https://github.com/phetsims/wave-interference/issues/316#issuecomment-454518891. Here, the only adjustment I made was a smooth slide from high to low separation. The lattice unexpectedly cleared, and the lattice ended up pretty noisy and took a bit of time to settle.

![noise4](https://user-images.githubusercontent.com/8419308/51283626-0cb58900-19a6-11e9-8aa5-741f74473732.gif)

I'm now leaning towards the first option proposed by @samreid in https://github.com/phetsims/wave-interference/issues/316#issuecomment-454458763:
> Clear the waves when the source separation changes (may also eliminate spurious artifacts, but makes it less user friendly when looking at a changing interference pattern in the wave area or on the screen since the user has to wait for re-propagation).

I think this would be somewhat similar to how we handle frequency changes for light, or how we clear the right side of the lattice when the barrier is moved.

**Comment by samreid** [comment]
I pushed a commit that suppresses waves while the separation is changing.  @arouinfar can you please review on master?

**Comment by samreid** [comment]
This version clears the sound particles.  Ready for testing
https://phet-dev.colorado.edu/html/wave-interference/1.0.0-dev.65/phet/wave-interference_en_phet.html

**Comment by arouinfar** [comment]
@samreid this is looking much cleaner to me. I like it!
